numba_dpex/core/kernel_interface/spirv_kernel.py

```diff
@@ -0,0 +1,175 @@
+# SPDX-FileCopyrightText: 2022 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from types import FunctionType
+
+from numba.core import compiler, ir
+from numba.core import types as numba_types
+from numba.core.compiler_lock import global_compiler_lock
+
+from numba_dpex import compiler as dpex_compiler
+from numba_dpex import config, spirv_generator
+from numba_dpex.core.descriptor import dpex_target
+from numba_dpex.core.exceptions import (
+    KernelHasReturnValueError,
+    UncompiledKernelError,
+    UnreachableError,
+)
+
+from .kernel_base import KernelInterface
+
+
+class SpirvKernel(KernelInterface):
+    def __init__(self, func, pyfunc_name) -> None:
+        self._llvm_module = None
+        self._device_driver_ir_module = None
+        self._module_name = None
+        self._pyfunc_name = pyfunc_name
+        self._func = func
+        if isinstance(func, FunctionType):
+            self._func_ty = FunctionType
+        elif isinstance(func, ir.FunctionIR):
+            self._func_ty = ir.FunctionIR
+        else:
+            raise UnreachableError()
+
+    @global_compiler_lock
+    def _compile(self, pyfunc, args, debug=None, extra_compile_flags=None):
+        """
+        Compiles the function using the dpex compiler pipeline and returns the
+        compiled result.
+
+        Args:
+            pyfunc: The function to be compiled. Can be a Python function or a
+            Numba IR object representing a function.
+            args: The list of arguments passed to the kernel.
+            debug (bool): Optional flag to turn on debug mode compilation.
+            extra_compile_flags: Extra flags passed to the compiler.
+
+        Returns:
+            cres: Compiled result.
+
+        Raises:
+            KernelHasReturnValueError: If the compiled function returns a
+            non-void value.
+        """
+        # First compilation will trigger the initialization of the backend.
+        typingctx = dpex_target.typing_context
+        targetctx = dpex_target.target_context
```

Comment:
@diptorupd What is the difference between this target context and the target context inside the function `compile()`?

Reply:
The contexts are now as function parameters to help make their usage cleaer.